File: leetecode/BTCameras.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// without memoization TLE
class Solution {
    
    int solve(TreeNode root,boolean cam,boolean mcam){
        if(root==null) return mcam ? 1001 : 0;
        int res = solve(root.left,true,false) + solve(root.right,true,false) + 1;
        if(mcam) return res;
        res = Math.min(solve(root.left,false,!cam)+solve(root.right,false,false),res);
        if(!cam) res = Math.min(solve(root.right,false,!cam)+solve(root.left,false,false),res);
        return res;
        
    }
    public int minCameraCover(TreeNode root) {
        if(root.left==null && root.right==null) return 1;
        return Math.min(solve(root,false,true),solve(root,false,false));
    }
}

File: leetecode/BTCameras.py

class Solution:
    def minCameraCover(self, root: TreeNode) -> int:
        def solve(root,cam,mcam):
            if root == None:
                return 1001 if mcam else 0
            res = solve(root.left,True,False) + solve(root.right,True,False) + 1
            if mcam:
                return res
            res = min(solve(root.left,False,not cam)+solve(root.right,False,False),res)
            if not cam:
                res = min(solve(root.right,False,not cam)+solve(root.left,False,False),res)
            return res
        if root.left == None and root.right == None:
            return 1
        return min(solve(root,False,True),solve(root,False,False))
    